### ISSUE TITLE
Alpine Linux build is broken

### ISSUE THREAD

**Issue by akoeplinger** [mono/mono#7167]
## Steps to Reproduce

1. Run build on Alpine Linux
2. The build runs very slowly during the `make` step (~4 hrs instead of 35mins) and then runs into all the timeouts of test steps so it's essentially broken

This started with https://jenkins.mono-project.com/job/test-mono-mainline-docker/label=docker-alpine/573/ which makes these commits suspect:

~1.  \[System\] System.Net.* tests from corefx (+RemoteExecutorTestBase) ([commit](https://github.com/mono/mono/commit/e9e17e069e6949600a382341e8888d7486553f21))~
~2.  IOS SDKs (# 6257) ([commit](https://github.com/mono/mono/commit/a2407ac76a50cc1b5fdfbea83adb135d9be40356))~
~3.  Remove a few braces that are contrary to our prescribed style. (# 6245) ([commit](https://github.com/mono/mono/commit/891941ecbe9640b180f73096b8b230c6f48bf11b))~
~4.  \[threads\] Ensure we remove the thread only once (# 6194) ([commit](https://github.com/mono/mono/commit/f6bf1d427b29ba00211f047081e492c56d06b41c))~
5.  \[threadpool\] Replace parked_threads condition variable by a semaphore ([commit](https://github.com/mono/mono/commit/5f5c5e97a08f7086d7c18af37352c4a03dc4c0d1))
~6.  \[runtime\] Fixes issue when compiling with -no\_weak\_imports (# 6276) ([commit](https://github.com/mono/mono/commit/6ca7b346cd3fc52318caccc182f35b8425439a07))~
~7.  Minor cleanup split out and expanded from another larger change. (# 6263) ([commit](https://github.com/mono/mono/commit/fba6bfc9549fb8686576cca20e33df4d01fb2d1a))~
~8.  This list is almost sorted -- so repair the sorting and add comment. ([commit](https://github.com/mono/mono/commit/11e18c2a5d6ed7d17184ad4d5e0606ab1e7ec334))~
~9.  Only call \`waitpid\` for processes that Mono tracks, rather than all ([commit](https://github.com/mono/mono/commit/8c9c37c45bb3302f917fe345004a5076c21d5ba1))~
~10.  cooperative suspend babystep wrt ([commit](https://github.com/mono/mono/commit/e0b1d6651915c395eca637ed172399354ee0fa17))~
~11.  \[System.Windows.Forms\] Fix CJK Encoding on Window Title ([commit](https://github.com/mono/mono/commit/c0e6ecb6c2ced241811e5569bfd483acd9cf6d94))~
~12.  \[mono-api-html\] do not warn about removal of \[Obsolete\] member. (#5294) ([commit](https://github.com/mono/mono/commit/05d5204da8f26549cf2e8ae387d907f6a4748823))~
~13.  \[System.Core\] Use corefx System.Linq tests (25k tests) (# 5515) ([commit](https://github.com/mono/mono/commit/cae8dd6dbd9253c0abbe2ab0a000ac617f116a78))~

I've crossed out those which I think are unrelated.

_Edit:_ **https://github.com/mono/mono/issues/7167#issuecomment-375898239 confirms 5f5c5e97 to be the issue**

## Current Behavior

Build is so slow that it is essentially broken.

## Expected Behavior

Not broken.

### On which platforms did you notice this

[ ] macOS
[ X ] Linux (Alpine)
[ ] Windows

**Version Used**: master

**Comment by jaykrell** [comment]
Wierd. Anything particularly different/wierd about this system? Anyway, can we bisect?? (I'm not practised in it, but I hear it is viable.)

**Comment by akoeplinger** [comment]
Yeah the "weirdest" thing is that it uses musl libc instead of glibc.

We can do bisect I just didn't bother with it yet if someone can tell the issue by looking at the commit :)

**Comment by jaykrell** [comment]
Presumably muslc related?

**Comment by jaykrell** [comment]
Yeah. I kinda think we should try to avoid libc and pthread dependencies on Linux and program the kernel interface. Or at least find a viable forkable renamable libc. Though this might be bad for interop reasons. Yeah I agree changing the C runtime could be hazardous.

**Comment by jaykrell** [comment]
No clear indication in jenkins if x86 or amd64, like if I setup a local VM.

**Comment by akoeplinger** [comment]
@jaykrell it's amd64. musl is used in Emscripten afaik so it could be important for the webassembly effort.

**Comment by akoeplinger** [comment]
Here are the packages we install on Jenkins if you want to setup a local environment:

```bash
$ apk update
$ apk add alpine-sdk wget ca-certificates tzdata bc git python linux-headers paxmark autoconf automake libtool gettext libxslt ttf-dejavu cmake lsof xvfb xauth
```

**Comment by jaykrell** [comment]
Do we really support such systems btw? It is hardly GNU/Linux without glibc. Yeah but webassembly doesn't have pthreads, right? :) Close enough proxy test? (all somewhat rhetorical questions -- "close enough" is subjective, webassembly is presumably a moving target and a moving aim..)

**Comment by jaykrell** [comment]
Seems to be on an old kernel too?

**Comment by akoeplinger** [comment]
well one thing I should note is that we're running this on Jenkins with Docker so the kernel would be from the underlying system - Ubuntu 16.04

**Comment by jaykrell** [comment]
also the hardening stuff?

**Comment by akoeplinger** [comment]
Not sure what you mean by that. It should be easy to reproduce since it worked normally for a few months before the mentioned commits.

**Comment by marek-safar** [comment]
/cc @luhenry 

Any idea what is so special on Alpine Linux?

**Comment by luhenry** [comment]
What are the steps to reproduce? I don't see the reason why it would only reproduce on Alpine Linux, but maybe some kernel of libc specific issues? I am shooting in the dark here, so being able to reproduce and possibly profile would definitely help.

**Comment by akoeplinger** [comment]
@luhenry just running the build is enough, it takes about 4hours on Jenkins now rather than 35minutes

**Comment by jaykrell** [comment]
For example, clean in mcs/class/corlib takes just over a minute in Alpine Linux VM and 11-12 seconds on native Mac. Alpine Linux has this pax/grsecurity stuff you have to disable to get this far, at least when using its kernel -- not sure about the docker situation. I wonder if there some emulation going on. Like maybe disabling security still takes NX faults on all JIT bytes or some other traps+emulation? Another experiment might be compare interpreter speed and see if there is slowdown there. The VM has 2GB of memory, 2 cores, just building one leaf. That could maybe explain a 5x slowdown vs. name. I can try giving it additional resources.

**Comment by jaykrell** [comment]
4 cores 4GB still takes over a minute.
I suspect it is either pax/grsecurity or less good pthread implementation but I don't have evidence just suspicion -- this really is a different system imho. Another thing to try would be statically link to glibc on another system and run that on AlpineLinux -- that'd give you same pthreads implementation. Or for that matter, build a muslc sysroot on Debian. Or AOT Rosyln and compare. Or maybe some more direct profiling or data gather -- like if the pax/grsecurity stuff has reportin.

**Comment by jaykrell** [comment]
Oh also something looked fishy with sigaltstack?

**Comment by jaykrell** [comment]
For sigaltstack comment, try running under a debugger

**Comment by awilfox** [comment]
Hi, I'm a developer of Adélie Linux, a derivative of Alpine.  Some of these comments are not very clear, and I'd like to answer some of your questions and ask my own:

1. No, there is no emulation of PaX if you are using the vanilla kernel.  (Adélie only ships our custom kernel, named `easy-kernel`, which follows the `mc` patchset and has no PaX or grsec.)
2. You're right, this isn't really a GNU/Linux system, but it is a Linux system.  The musl libc is typically a lot more performant than glibc, and it's really odd that there is a performance regression here.  For example, Firefox starts about 40% faster on Adélie than it does on Fedora, and uses about 15% less memory.  This is mostly because musl's memory allocator uses less resources, but is also because their threading implementation is lighter and simpler.

Once our alpha5 release packaging is done I would like to play around with Mono and see if I can narrow it down further.

Are these builds being run under Docker with another OS as the host?  If that is the case, then you can definitely count PaX and grsec out unless your host OS has that enabled.  That is a kernel-only security package and it definitely doesn't affect any of the userland.

Are there any special steps for building beyond the documented method that I need to do to make this problem more evident?  Some `configure` flag that makes it worse?

Regarding sigaltstack: I saw that this option turns stack overflows into StackOverflowException.  By default, musl threads have just 80K stack space.  Does Mono manually set the stack space to a sane value (as required by POSIX, since the stack space is not guaranteed)?  A call to [pthread_attr_setstacksize](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_getstacksize.html) should suffice.  If not, this could be the cause of the slowdowns.

**Comment by marek-safar** [comment]
I don't think we run the build under Docker but it's virtualized environment.

The build should be simple `autogen.sh` then `make` when you have all dependencies installed. For detailed instructions see http://www.mono-project.com/docs/compiling-mono/linux/

We do call/set `setstacksize` but the managed build is quite complicated so it's possible we set large values than the system expects

**Comment by akoeplinger** [comment]
@marek-safar we do indeed run the Alpine build in Docker on Ubuntu 16.04 since it's the easiest way to get it going :)

**Comment by jaykrell** [comment]
We might try Adelie instead of Alpine as well. That would be one of the quicker experiments -- short of debugging. I ran into the pax/grsecurity stuff because I do have Alpine in a "real" VM, not docker. If we can rule out pax/grsecurity as kernel only and not relevant then muslc is the main suspect -- but still speculation, no debugging.

Just using "regular" Linux and Windows I've seen code sensitive to and requiring a quite large stack, like gcc requires 8MB I recall. But maybe muslc is only small for new threads, not the first thread, for increased compatibilty? Stack and its size is a tough thing, no great answers.

We should also try mcs.

**Comment by jaykrell** [comment]
Bisect gave me the wrong answer, trying again.

**Comment by jaykrell** [comment]
Bisect says it is https://github.com/mono/mono/commit/5f5c5e97a08f7086d7c18af37352c4a03dc4c0d1

[threadpool] Replace parked_threads condition variable by a semaphore #6222

Later to confirm with master and reverting it and then to try to understand it..

**Comment by jaykrell** [comment]
Confirmed on Alpine with revert.
Also easy to reproduce on Debian 9 with muslc (with cross compiler form).

**Comment by luhenry** [comment]
Could you provide a performance profile comparison between both implementation? Could you also run it on various platforms? What's the reason for the semaphore implementation to be so slow on Alpine Linux and it doesn't seem to be the problem on any other platform? Isn't it something that should instead be reported against Alpine Linux?

**Comment by jaykrell** [comment]
Also what other platforms? I could try other architectures or other distributions but Debian/amd64/musl would seem sufficient.

**Comment by luhenry** [comment]
Semaphores in the case of the threadpool solves a whole lot of issues over condition variables, so we will need to work around musl, especially since it only affects this specific configuration.

**Comment by jaykrell** [comment]
Can't you trivially build a semaphore out of a condition variable?

Condition variables seem to easily efficiently build all kinds of things. I find they provide an easier programming model in every case. With semaphores (or windows events) you have to bottleneck all your interesting state to one or a small number of integers or booleans. With condition variables you just write an arbitrary expression instead.

Take a lock, increment or decrement an integer, and signal the condition variable for any of the transitions you care about -- all of them, or more likely from one to zero, or from zero to one.

Does the semaphore provide for less lock contention, i.e. for all the non-signaled transitions?

**Comment by luhenry** [comment]
If you think that is the fix for musl, then feel free to submit a PR that implements a semaphore on top of a condition variable. I don't see why only the threadpool would be impacted in this case. Just note that since this slow semaphore issue only affects musl, then the semaphore on top of the condition variable should only be implemented for musl, as a workaround.

You can find the wrapper we use for semaphores at https://github.com/mono/mono/blob/master/mono/utils/mono-os-semaphore.h

**Comment by jaykrell** [comment]
I wonder if the musl semaphores are never woken except by time out. We should experiment..

**Comment by jaykrell** [comment]
I agree I don't see why threadpool only is impacted. And it isn't clearly.
Everything works albeit slowly. Maybe threadpool is only most impacted.

**Comment by jaykrell** [comment]
We could ifdef muslc go back to the old way, but I'd kind not like to do that. Our binaries should avoid being glibc or muslc-specific. One set of binaries is ideally very portable across multiple Linux distributions/distributions -- assuming glibc and muslc are very very binary compatible.

We could go back to the old way?
Did it work ok?

There is some third way that works well on both/all?

I also wouldn't really like a runtime determination here.

If we just ifdef muslc or runtime check, that leaves a dormant code path. Better to have fewer code paths with more testing.

**Comment by frebib** [comment]
Reverting 5f5c5e9 took our build from 3:30-5:30h to ~20 minutes. I'm hoping reverting the commit doesn't cause any issues until this is fixed

**Comment by rootwyrm** [comment]
`alpine:edge` has ceased to reproduce this issue on all versions of mono including 5.14, 5.18, 5.20, and head. It still reproduces in 3.8 and 3.9. I can't find any clearly corresponding bugfixes or commits to explain it, which is unfortunately, indicative of exactly nothing. However, the behavior change seems to indicate the presence of an upstream regression that is resolved. 

Building with docker hub on `edge`:
```
May 15 13:27:25 [MONO] Configuring...
<omitted>
May 15 14:42:29 [MONO] Build and install of 5.20.1.27 complete.
```

This is actually much better than previous; 5.8 and 5.10 clocked in at around 2-3 hours on Hub. This has it down to 75 minutes. When using `alpine:3.9` it still consistently hits the 8 hour dead-hang and is killed. The fix appears to be somewhere between `musl-1.1.22` and `1.1.20`, or the addition of `libc6-compat` from musl.

**Comment by jaykrell** [comment]
Great! Data for ubuntu/glibc etc.?

**Comment by frebib** [comment]
A few changes from https://git.musl-libc.org/cgit/musl/tree/WHATSNEW that seem relevant:
line 1960 - some futex waits omitted timeout arg to syscall, thereby spun on EFAULT
line 2033 - sem_wait and sem_timedwait were wrongly not interruptible by signals

I'm guessing it was probably that second change, but it's still great to see that this is improved now :tada:

**Comment by rootwyrm** [comment]
Based on what little analysis I was able to perform (Alpine is way too stripped down in some regards,) I would actually pin it on 1960 as I did observe constant FUTEX spinning and failing to timeout on broken builds. But that's only slightly informed speculation here, due to issues with instrumenting the build.

I don't have an Ubuntu environment available, but I do have a CentOS 7.6 (used for testing the Docker builds in fact) so I can test same-hardware no problem.

CentOS 7.6, Docker, [rootwyrm/tc_mono:latest](https://github.com/rootwyrm/talecaster/tree/master/docker/tc_mono):
```
$ time docker build . -t mono_timed_520127
(output omitted)
May 15 21:48:26 [MONO] Build and install of 5.20.1.27 complete.

real 30m33.327s
user 0m0.422s
sys  0m0.657s
```
So yes, we're very decently fast here. Actually under 25 minutes from spawning the initial container to functional binaries, since total runtime includes running hooks, updating external sources, downloading the mono source, and installing a number of supporting packages. 

CentOS 7.6 bare, same system
```
$ time ./configure --prefix=/usr/local --sysconfdir=/usr/local/etc --mandir=/usr/share/man --enable-ninja --enable-system-aot --disable-boehm --without-x --with-mcs-docs=no --enable-small-config
(output omitted)
real    0m30.693s
user    0m14.219s
sys     0m17.645s
$ time make
(output omitted)
real    25m7.074s
user    29m30.710s
sys     4m14.581s
```
As we can see here, the actual compile time for glibc is edging toward _slower_ than musl, and very likely is slower than musl when eliminating the Docker overhead.

**Comment by awilfox** [comment]
If you need a musl environment that is less stripped bare, consider [Adélie Linux](https://www.adelielinux.org/); we don't have a Docker image yet, but we do have -dbg packages for everything we ship (except LLVM, due to issues handling debugging symbols on binaries that size).

**Comment by jaykrell** [comment]
I had a decent environment a year ago.
I don't remember if it was Alpine or Adelie, and muslc on Debian or Ubuntu.
I went back later to recreate the muslc on Debian or Ubuntu and couldn't get it to work easily.
Very recently I was going to try again and #ifndef glibc but it seems kinda not needed now.

and non interruptibability would be the thing.

We have a random finite sleep in the threadpool but we interrupt it at shutdown.
The hang is that sleep expiring w/o interruption. I think.

Docker didn't work out well on my Mac at the time so I gave up.
I should try again soon though if I have time.

**Comment by NattyNarwhal** [comment]
I don't notice any performance issues with musl (using mcs though) anymore.

**Comment by akoeplinger** [comment]
Looks like csc build is completely broken now: https://jenkins.mono-project.com/job/test-mono-mainline-docker/label=docker-alpine/

**Comment by rootwyrm** [comment]
Looking into this. I have seen similar problems before with certain configuration flags, but this does look different from that. 6.0 and 6.4 still seem fine.

**Comment by rootwyrm** [comment]
This is an interesting one.. things are fine until it hits CSC, and then it's the same hanging in `FUTEX` behavior observed on musl before. However, it's definitely not musl this time, because it's reproducing on Alpine 3.9, 3.10, 3.10.3, and edge alike. On multiple differing hosts, too; got it to produce on both Ubuntu and CentOS.

```
mmap(0x41d09000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d09000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7eff8b015000
futex(0x556b4f343a48, FUTEX_WAKE_PRIVATE, 1) = 1
mmap(0x41d19000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d19000
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
futex(0x7ffdd91fd6a4, FUTEX_WAIT_PRIVATE, 2, NULL) = 0
munmap(0x7eff8b056000, 4096)            = 0
madvise(0x556b5108b000, 520192, MADV_DONTNEED) = 0
madvise(0x556b5108b000, 520192, MADV_DONTNEED) = 0
madvise(0x556b5108b000, 520192, MADV_DONTNEED) = 0
madvise(0x556b5108b000, 520192, MADV_DONTNEED) = 0
madvise(0x556b50b23000, 1064960, MADV_DONTNEED) = 0
madvise(0x556b50b27000, 1048576, MADV_DONTNEED) = 0
getpid()                                = 52352
sched_getaffinity(52352, 128, [0 1])    = 128
mmap(0x41d29000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d29000
getpid()                                = 52352
sched_getaffinity(52352, 128, [0 1])    = 128
getpid()                                = 52352
sched_getaffinity(52352, 128, [0 1])    = 128
mmap(0x41d39000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d39000
madvise(0x556b50fde000, 266240, MADV_DONTNEED) = 0
mmap(0x41d49000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d49000
madvise(0x556b50fde000, 266240, MADV_DONTNEED) = 0
madvise(0x556b50fde000, 266240, MADV_DONTNEED) = 0
madvise(0x556b50fde000, 266240, MADV_DONTNEED) = 0
madvise(0x556b50fde000, 266240, MADV_DONTNEED) = 0
madvise(0x556b5126f000, 163840, MADV_DONTNEED) = 0
mmap(0x41d59000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d59000
madvise(0x556b50bff000, 163840, MADV_DONTNEED) = 0
madvise(0x556b50be8000, 258048, MADV_DONTNEED) = 0
madvise(0x556b50b93000, 606208, MADV_DONTNEED) = 0
mmap(0x41d69000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d69000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7eff8b056000
mmap(0x41d79000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d79000
madvise(0x556b50ba6000, 528384, MADV_DONTNEED) = 0
madvise(0x556b51214000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
mmap(0x41d89000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d89000
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be7000, 262144, MADV_DONTNEED) = 0
madvise(0x556b50be8000, 258048, MADV_DONTNEED) = 0
madvise(0x556b50be8000, 258048, MADV_DONTNEED) = 0
madvise(0x556b50be8000, 258048, MADV_DONTNEED) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7eff8b014000
mmap(0x41d99000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41d99000
mmap(0x41da9000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41da9000
madvise(0x556b512ff000, 266240, MADV_DONTNEED) = 0
mmap(0x41db9000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41db9000
madvise(0x556b5133b000, 266240, MADV_DONTNEED) = 0
madvise(0x556b5133b000, 266240, MADV_DONTNEED) = 0
mmap(0x41dc9000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41dc9000
mmap(0x41dd9000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41387000
getpid()                                = 52352
sched_getaffinity(52352, 128, [0 1])    = 128
mmap(0x41397000, 65536, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS|MAP_32BIT, -1, 0) = 0x41397000
futex(0x556b4f343a48, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x556b4f33f4c4, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x556b4f33f4c4, FUTEX_LOCK_PI_PRIVATE, 0, NULL) = 0
futex(0x556b4f33f4c4, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x556b4f33f4c4, FUTEX_LOCK_PI_PRIVATE, 0, NULL) = 0
futex(0x556b4f33f4c4, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x556b4f33f4c4, FUTEX_LOCK_PI_PRIVATE, 0, NULL) = 0
futex(0x556b4f33f4c4, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x556b4f33f4c4, FUTEX_LOCK_PI_PRIVATE, 0, NULL) = 0
futex(0x556b4f33f4c4, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x7eff887bb914, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x556b4f33ce4c, FUTEX_UNLOCK_PI_PRIVATE, 2147536000) = 0
futex(0x556b4f33ce4c, FUTEX_LOCK_PI_PRIVATE, 0, NULL) = -1 EINVAL (Invalid argument)
futex(0x7ffdd91fcdfc, FUTEX_WAIT_PRIVATE, 0, NULL
```

Never progresses that point, just hangs in that `FUTEX_WAIT_PRIVATE` again.

**Comment by rickardp** [comment]
What is the status of the Alpine build? It looks like it has been broken on Jenkins for a while.

**Comment by rickardp** [comment]
Got the same error here trying to build from source now. Got stuck in gensources.exe.

Backtrace:
```

#0  __syscall_cp_c (nr=202, u=140730101368332, v=128, w=0, x=0, y=0, z=0) at ./arch/x86_64/syscall_arch.h:61
#1  0x00007f34c9774257 in __futex4_cp (to=0x0, val=0, op=128, addr=0x7ffe47b3860c) at src/thread/__timedwait.c:52
#2  __timedwait_cp (addr=addr@entry=0x7ffe47b3860c, val=val@entry=0, clk=clk@entry=0, at=at@entry=0x0, priv=128, priv@entry=1) at src/thread/__timedwait.c:52
#3  0x00007f34c97742fe in __timedwait (addr=addr@entry=0x7ffe47b3860c, val=val@entry=0, clk=clk@entry=0, at=at@entry=0x0, priv=priv@entry=1) at src/thread/__timedwait.c:68
#4  0x00007f34c9776285 in pthread_mutex_timedlock_pi (at=0x0, m=<optimized out>) at src/thread/pthread_mutex_timedlock.c:51
#5  __pthread_mutex_timedlock (m=<optimized out>, at=at@entry=0x0) at src/thread/pthread_mutex_timedlock.c:68
#6  0x00007f34c9776122 in __pthread_mutex_lock (m=m@entry=0x55d4d5a59d68 <compilation_data+8>) at src/thread/pthread_mutex_lock.c:9
#7  0x000055d4d55fe109 in mono_os_mutex_lock (mutex=0x55d4d5a59d68 <compilation_data+8>) at ../../mono/utils/mono-os-mutex.h:103
#8  mono_coop_mutex_lock (mutex=0x55d4d5a59d68 <compilation_data+8>) at ../../mono/utils/mono-coop-mutex.h:54
#9  lock_compilation_data () at mini-runtime.c:1961
#10 unregister_method_for_compile (target_domain=0x55d4d6fe19e0, method=0x55d4d7ef4698) at mini-runtime.c:2087
#11 mono_jit_compile_method_with_opt (method=method@entry=0x55d4d7ef4698, opt=<optimized out>, jit_only=jit_only@entry=0, error=error@entry=0x7ffe47b387f0) at mini-runtime.c:2445
#12 0x000055d4d55fe930 in mono_jit_compile_method (method=method@entry=0x55d4d7ef4698, error=error@entry=0x7ffe47b387f0) at mini-runtime.c:2494
#13 0x000055d4d56346c6 in common_call_trampoline (regs=regs@entry=0x7ffe47b388c0, code=code@entry=0x40d2823c "H\213\365H\201\306P\377\377\377H\213\275\350\376\377\377\272\060", m=m@entry=0x55d4d7ef4698,
    vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, error=error@entry=0x7ffe47b387f0) at mini-trampolines.c:635
#14 0x000055d4d5635240 in mono_magic_trampoline (regs=0x7ffe47b388c0, code=0x40d2823c "H\213\365H\201\306P\377\377\377H\213\275\350\376\377\377\272\060", arg=0x55d4d7ef4698, tramp=<optimized out>)

```

**Comment by rickardp** [comment]
Had some spare time today so I did a git bisect, and this commit is the culprit:
3e8108ea6576b07de2a64528be18674683879189
Easily reproducible in a Docker file

**Comment by rickardp** [comment]
I created PR #19914 that fixes the broken Alpine build. I have not yet checked for other issues on Alpine as I've just recently got Mono to build and test OK.

**Comment by rootwyrm** [comment]
Wow, thanks for the excellent catch on this @rickardp ! That would indeed seem to be the problem.
